File: notebook/collab/locks.py
"""Cell locking mechanism for collaborative notebook editing.

This module implements the server-side component of the cell locking mechanism,
preventing editing conflicts by allowing only one user to edit a cell at a time.
It manages lock acquisition, release, timeout handling, and conflict resolution.
"""

import asyncio
import json
import logging
import time
from datetime import datetime, timedelta
from typing import Dict, List, Optional, Set, Tuple, Any

from tornado.ioloop import IOLoop, PeriodicCallback
from tornado.websocket import WebSocketHandler

from notebook.collab.persistence import CollaborationPersistence
from notebook.collab.permissions import PermissionManager

# Configure logger
logger = logging.getLogger(__name__)

# Lock timeout in seconds (default: 5 minutes)
DEFAULT_LOCK_TIMEOUT = 300

# Lock cleanup interval in milliseconds (default: 30 seconds)
LOCK_CLEANUP_INTERVAL = 30000

# Lock types
class LockType:
    """Enumeration of lock types."""
    EDIT = "edit"  # Lock for editing a cell
    EXECUTE = "execute"  # Lock for executing a cell
    METADATA = "metadata"  # Lock for changing cell metadata


class LockStatus:
    """Enumeration of lock statuses."""
    ACQUIRED = "acquired"  # Lock successfully acquired
    DENIED = "denied"  # Lock acquisition denied
    RELEASED = "released"  # Lock successfully released
    EXPIRED = "expired"  # Lock expired due to timeout
    OVERRIDE = "override"  # Lock forcibly overridden by admin


class Lock:
    """Represents a cell lock in a collaborative notebook."""
    
    def __init__(self, 
                 document_id: str, 
                 cell_id: str, 
                 user_id: str, 
                 lock_type: str = LockType.EDIT,
                 timeout: int = DEFAULT_LOCK_TIMEOUT):
        """Initialize a new lock.
        
        Args:
            document_id: ID of the notebook document
            cell_id: ID of the cell being locked
            user_id: ID of the user acquiring the lock
            lock_type: Type of lock (edit, execute, metadata)
            timeout: Lock timeout in seconds
        """
        self.document_id = document_id
        self.cell_id = cell_id
        self.user_id = user_id
        self.lock_type = lock_type
        self.acquired_at = time.time()
        self.expires_at = self.acquired_at + timeout
        self.timeout = timeout
        self.last_activity = self.acquired_at
        
    def is_expired(self) -> bool:
        """Check if the lock has expired.
        
        Returns:
            True if the lock has expired, False otherwise
        """
        return time.time() > self.expires_at
    
    def update_activity(self) -> None:
        """Update the last activity timestamp and extend the expiration time."""
        self.last_activity = time.time()
        self.expires_at = self.last_activity + self.timeout
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert the lock to a dictionary for serialization.
        
        Returns:
            Dictionary representation of the lock
        """
        return {
            "document_id": self.document_id,
            "cell_id": self.cell_id,
            "user_id": self.user_id,
            "lock_type": self.lock_type,
            "acquired_at": self.acquired_at,
            "expires_at": self.expires_at,
            "last_activity": self.last_activity,
            "timeout": self.timeout
        }
    
    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'Lock':
        """Create a lock from a dictionary.
        
        Args:
            data: Dictionary representation of a lock
            
        Returns:
            A new Lock instance
        """
        lock = cls(
            document_id=data["document_id"],
            cell_id=data["cell_id"],
            user_id=data["user_id"],
            lock_type=data["lock_type"],
            timeout=data["timeout"]
        )
        lock.acquired_at = data["acquired_at"]
        lock.expires_at = data["expires_at"]
        lock.last_activity = data["last_activity"]
        return lock


class LockManager:
    """Manages cell locks for collaborative notebook editing."""
    
    def __init__(self, persistence: CollaborationPersistence, permission_manager: PermissionManager):
        """Initialize the lock manager.
        
        Args:
            persistence: Persistence layer for storing locks
            permission_manager: Permission manager for checking user permissions
        """
        self.persistence = persistence
        self.permission_manager = permission_manager
        
        # In-memory cache of active locks: {document_id: {cell_id: Lock}}
        self._locks: Dict[str, Dict[str, Lock]] = {}
        
        # Set of connected clients by user_id
        self._connected_users: Set[str] = set()
        
        # Start periodic cleanup of expired locks
        self._cleanup_task = PeriodicCallback(
            self._cleanup_expired_locks,
            LOCK_CLEANUP_INTERVAL
        )
        self._cleanup_task.start()
        
        # Load existing locks from persistence
        IOLoop.current().add_callback(self._load_locks)
    
    async def _load_locks(self) -> None:
        """Load existing locks from persistence layer."""
        try:
            locks_data = await self.persistence.get_all_locks()
            for lock_data in locks_data:
                lock = Lock.from_dict(lock_data)
                if not lock.is_expired():
                    self._add_lock_to_cache(lock)
            logger.info(f"Loaded {len(locks_data)} locks from persistence")
        except Exception as e:
            logger.error(f"Error loading locks from persistence: {e}")
    
    def _add_lock_to_cache(self, lock: Lock) -> None:
        """Add a lock to the in-memory cache.
        
        Args:
            lock: The lock to add
        """
        if lock.document_id not in self._locks:
            self._locks[lock.document_id] = {}
        self._locks[lock.document_id][lock.cell_id] = lock
    
    def _remove_lock_from_cache(self, document_id: str, cell_id: str) -> None:
        """Remove a lock from the in-memory cache.
        
        Args:
            document_id: ID of the notebook document
            cell_id: ID of the cell
        """
        if document_id in self._locks and cell_id in self._locks[document_id]:
            del self._locks[document_id][cell_id]
            if not self._locks[document_id]:
                del self._locks[document_id]
    
    async def acquire_lock(self, 
                          document_id: str, 
                          cell_id: str, 
                          user_id: str, 
                          lock_type: str = LockType.EDIT,
                          force: bool = False) -> Tuple[str, Optional[Lock]]:
        """Attempt to acquire a lock on a cell.
        
        Args:
            document_id: ID of the notebook document
            cell_id: ID of the cell to lock
            user_id: ID of the user requesting the lock
            lock_type: Type of lock (edit, execute, metadata)
            force: Whether to force acquire the lock (admin override)
            
        Returns:
            Tuple of (status, lock) where status is one of the LockStatus values
            and lock is the acquired lock or None if acquisition failed
        """
        # Check if user has permission to edit this cell
        has_permission = await self.permission_manager.check_cell_permission(
            document_id, cell_id, user_id, "edit"
        )
        
        if not has_permission and not force:
            logger.warning(f"User {user_id} denied lock on {document_id}/{cell_id} due to permissions")
            return LockStatus.DENIED, None
        
        # Check if cell is already locked
        existing_lock = self._get_lock(document_id, cell_id)
        
        # If there's an existing lock
        if existing_lock:
            # If it's the same user, just update the activity timestamp
            if existing_lock.user_id == user_id:
                existing_lock.update_activity()
                await self.persistence.update_lock(existing_lock.to_dict())
                return LockStatus.ACQUIRED, existing_lock
            
            # If it's expired, release it and continue with acquisition
            if existing_lock.is_expired():
                await self.release_lock(document_id, cell_id, existing_lock.user_id, expired=True)
            # If force flag is set and user has admin permission, override the lock
            elif force:
                is_admin = await self.permission_manager.check_document_permission(
                    document_id, user_id, "admin"
                )
                if is_admin:
                    await self.release_lock(document_id, cell_id, existing_lock.user_id, override=True, override_by=user_id)
                else:
                    logger.warning(f"User {user_id} attempted to force acquire lock without admin permission")
                    return LockStatus.DENIED, None
            else:
                # Lock is valid and held by another user
                logger.info(f"Lock acquisition denied for {user_id} on {document_id}/{cell_id}, already locked by {existing_lock.user_id}")
                return LockStatus.DENIED, None
        
        # Create and store the new lock
        lock = Lock(document_id, cell_id, user_id, lock_type)
        self._add_lock_to_cache(lock)
        await self.persistence.add_lock(lock.to_dict())
        
        logger.info(f"Lock acquired for {user_id} on {document_id}/{cell_id}")
        return LockStatus.ACQUIRED, lock
    
    async def release_lock(self, 
                          document_id: str, 
                          cell_id: str, 
                          user_id: str, 
                          expired: bool = False,
                          override: bool = False,
                          override_by: Optional[str] = None) -> Tuple[str, Optional[Dict[str, Any]]]:
        """Release a lock on a cell.
        
        Args:
            document_id: ID of the notebook document
            cell_id: ID of the cell
            user_id: ID of the user releasing the lock
            expired: Whether the lock is being released due to expiration
            override: Whether the lock is being forcibly released (admin override)
            override_by: ID of the admin user overriding the lock
            
        Returns:
            Tuple of (status, lock_info) where status is one of the LockStatus values
            and lock_info contains information about the released lock
        """
        lock = self._get_lock(document_id, cell_id)
        
        if not lock:
            logger.warning(f"Attempted to release non-existent lock on {document_id}/{cell_id}")
            return LockStatus.RELEASED, None
        
        # Only the lock owner or an admin can release a lock
        if lock.user_id != user_id and not override:
            is_admin = await self.permission_manager.check_document_permission(
                document_id, user_id, "admin"
            )
            if not is_admin:
                logger.warning(f"User {user_id} denied lock release on {document_id}/{cell_id}, owned by {lock.user_id}")
                return LockStatus.DENIED, None
        
        # Prepare lock info for notification
        lock_info = {
            "document_id": document_id,
            "cell_id": cell_id,
            "user_id": lock.user_id,
            "lock_type": lock.lock_type,
            "released_by": user_id,
            "expired": expired,
            "override": override,
            "override_by": override_by
        }
        
        # Remove the lock
        self._remove_lock_from_cache(document_id, cell_id)
        await self.persistence.remove_lock(document_id, cell_id)
        
        status = LockStatus.EXPIRED if expired else LockStatus.OVERRIDE if override else LockStatus.RELEASED
        logger.info(f"Lock released for {lock.user_id} on {document_id}/{cell_id} (status: {status})")
        
        return status, lock_info
    
    def _get_lock(self, document_id: str, cell_id: str) -> Optional[Lock]:
        """Get the current lock for a cell if it exists.
        
        Args:
            document_id: ID of the notebook document
            cell_id: ID of the cell
            
        Returns:
            The Lock object if found, None otherwise
        """
        if document_id in self._locks and cell_id in self._locks[document_id]:
            return self._locks[document_id][cell_id]
        return None
    
    async def get_document_locks(self, document_id: str) -> List[Dict[str, Any]]:
        """Get all locks for a document.
        
        Args:
            document_id: ID of the notebook document
            
        Returns:
            List of lock dictionaries for the document
        """
        if document_id not in self._locks:
            return []
        
        return [lock.to_dict() for lock in self._locks[document_id].values()]
    
    async def check_lock(self, document_id: str, cell_id: str, user_id: str) -> Tuple[bool, Optional[Dict[str, Any]]]:
        """Check if a user can edit a cell.
        
        Args:
            document_id: ID of the notebook document
            cell_id: ID of the cell
            user_id: ID of the user
            
        Returns:
            Tuple of (can_edit, lock_info) where can_edit is True if the user can edit the cell
            and lock_info contains information about any existing lock
        """
        lock = self._get_lock(document_id, cell_id)
        
        # No lock exists, user can edit
        if not lock:
            return True, None
        
        # Lock exists but is expired, clean it up and allow edit
        if lock.is_expired():
            status, lock_info = await self.release_lock(document_id, cell_id, lock.user_id, expired=True)
            return True, lock_info
        
        # User holds the lock, can edit
        if lock.user_id == user_id:
            lock.update_activity()
            await self.persistence.update_lock(lock.to_dict())
            return True, lock.to_dict()
        
        # Another user holds the lock, cannot edit
        return False, lock.to_dict()
    
    async def _cleanup_expired_locks(self) -> None:
        """Periodically clean up expired locks."""
        try:
            documents_to_check = list(self._locks.keys())
            for document_id in documents_to_check:
                if document_id not in self._locks:
                    continue
                    
                cells_to_check = list(self._locks[document_id].keys())
                for cell_id in cells_to_check:
                    if cell_id not in self._locks[document_id]:
                        continue
                        
                    lock = self._locks[document_id][cell_id]
                    if lock.is_expired():
                        logger.info(f"Cleaning up expired lock for {lock.user_id} on {document_id}/{cell_id}")
                        await self.release_lock(document_id, cell_id, lock.user_id, expired=True)
        except Exception as e:
            logger.error(f"Error during lock cleanup: {e}")
    
    async def user_connected(self, user_id: str) -> None:
        """Track a user connection.
        
        Args:
            user_id: ID of the connected user
        """
        self._connected_users.add(user_id)
    
    async def user_disconnected(self, user_id: str) -> List[Dict[str, Any]]:
        """Handle user disconnection by releasing their locks.
        
        Args:
            user_id: ID of the disconnected user
            
        Returns:
            List of released lock information
        """
        self._connected_users.discard(user_id)
        
        # Find all locks held by this user
        released_locks = []
        documents_to_check = list(self._locks.keys())
        
        for document_id in documents_to_check:
            if document_id not in self._locks:
                continue
                
            cells_to_check = list(self._locks[document_id].keys())
            for cell_id in cells_to_check:
                if cell_id not in self._locks[document_id]:
                    continue
                    
                lock = self._locks[document_id][cell_id]
                if lock.user_id == user_id:
                    status, lock_info = await self.release_lock(
                        document_id, cell_id, user_id, expired=True
                    )
                    if lock_info:
                        released_locks.append(lock_info)
        
        if released_locks:
            logger.info(f"Released {len(released_locks)} locks for disconnected user {user_id}")
        
        return released_locks


class LockHandler:
    """Handles lock-related WebSocket messages."""
    
    def __init__(self, lock_manager: LockManager):
        """Initialize the lock handler.
        
        Args:
            lock_manager: The lock manager instance
        """
        self.lock_manager = lock_manager
    
    async def handle_message(self, message: Dict[str, Any], handler: WebSocketHandler) -> Optional[Dict[str, Any]]:
        """Handle a lock-related WebSocket message.
        
        Args:
            message: The message to handle
            handler: The WebSocket handler instance
            
        Returns:
            Response message to send back to the client, or None if no response is needed
        """
        action = message.get("action")
        document_id = message.get("document_id")
        cell_id = message.get("cell_id")
        user_id = message.get("user_id")
        lock_type = message.get("lock_type", LockType.EDIT)
        force = message.get("force", False)
        
        if not all([action, document_id, user_id]):
            return {
                "type": "lock_error",
                "error": "Missing required fields",
                "message_id": message.get("message_id")
            }
        
        if action == "acquire":
            if not cell_id:
                return {
                    "type": "lock_error",
                    "error": "Missing cell_id for lock acquisition",
                    "message_id": message.get("message_id")
                }
                
            status, lock = await self.lock_manager.acquire_lock(
                document_id, cell_id, user_id, lock_type, force
            )
            
            response = {
                "type": "lock_response",
                "action": "acquire",
                "status": status,
                "document_id": document_id,
                "cell_id": cell_id,
                "user_id": user_id,
                "message_id": message.get("message_id")
            }
            
            if lock:
                response["lock"] = lock.to_dict()
            
            return response
            
        elif action == "release":
            if not cell_id:
                return {
                    "type": "lock_error",
                    "error": "Missing cell_id for lock release",
                    "message_id": message.get("message_id")
                }
                
            status, lock_info = await self.lock_manager.release_lock(
                document_id, cell_id, user_id, 
                override=force, 
                override_by=user_id if force else None
            )
            
            response = {
                "type": "lock_response",
                "action": "release",
                "status": status,
                "document_id": document_id,
                "cell_id": cell_id,
                "user_id": user_id,
                "message_id": message.get("message_id")
            }
            
            if lock_info:
                response["lock_info"] = lock_info
            
            return response
            
        elif action == "check":
            if not cell_id:
                return {
                    "type": "lock_error",
                    "error": "Missing cell_id for lock check",
                    "message_id": message.get("message_id")
                }
                
            can_edit, lock_info = await self.lock_manager.check_lock(
                document_id, cell_id, user_id
            )
            
            response = {
                "type": "lock_response",
                "action": "check",
                "can_edit": can_edit,
                "document_id": document_id,
                "cell_id": cell_id,
                "user_id": user_id,
                "message_id": message.get("message_id")
            }
            
            if lock_info:
                response["lock_info"] = lock_info
            
            return response
            
        elif action == "list":
            locks = await self.lock_manager.get_document_locks(document_id)
            
            return {
                "type": "lock_response",
                "action": "list",
                "document_id": document_id,
                "locks": locks,
                "message_id": message.get("message_id")
            }
            
        else:
            return {
                "type": "lock_error",
                "error": f"Unknown lock action: {action}",
                "message_id": message.get("message_id")
            }
    
    async def handle_connection(self, user_id: str) -> None:
        """Handle a new WebSocket connection.
        
        Args:
            user_id: ID of the connected user
        """
        await self.lock_manager.user_connected(user_id)
    
    async def handle_disconnection(self, user_id: str) -> List[Dict[str, Any]]:
        """Handle a WebSocket disconnection.
        
        Args:
            user_id: ID of the disconnected user
            
        Returns:
            List of released lock information
        """
        return await self.lock_manager.user_disconnected(user_id)

File: packages/notebook/src/collab/locks.ts
/**
 * Cell locking mechanism for collaborative notebooks
 *
 * This module implements a cell locking system that prevents editing conflicts
 * by ensuring that only one user can edit a cell at a time. It manages the
 * acquisition and release of locks, broadcasts lock state to all connected clients,
 * and provides visual indicators for locked cells.
 */

import { EventEmitter } from '@lumino/signaling';
import * as Y from 'yjs';

import { IPermissionManager, UserPermission } from './permissions';

/**
 * Interface for cell lock data
 */
export interface ICellLock {
  /** ID of the locked cell */
  cellId: string;
  /** ID of the user who owns the lock */
  userId: string;
  /** User display name for UI */
  userName: string;
  /** Timestamp when the lock was acquired */
  timestamp: number;
  /** Optional timeout in milliseconds (0 means no timeout) */
  timeout: number;
}

/**
 * Lock event types
 */
export enum LockEventType {
  /** Lock was acquired */
  Acquired = 'acquired',
  /** Lock was released */
  Released = 'released',
  /** Lock timed out */
  TimedOut = 'timedout',
  /** Lock was forcibly released by admin */
  ForceReleased = 'forcereleased'
}

/**
 * Interface for lock events
 */
export interface ICellLockEvent {
  /** Type of lock event */
  type: LockEventType;
  /** The lock data */
  lock: ICellLock;
}

/**
 * Options for the CellLockManager
 */
export interface ICellLockManagerOptions {
  /** The Yjs document */
  doc: Y.Doc;
  /** The permission manager for access control */
  permissionManager: IPermissionManager;
  /** Default lock timeout in milliseconds (default: 5 minutes) */
  defaultTimeout?: number;
  /** Cleanup interval for checking expired locks in milliseconds (default: 30 seconds) */
  cleanupInterval?: number;
}

/**
 * Manager for cell locks in collaborative notebooks
 */
export class CellLockManager {
  /**
   * Constructor
   * 
   * @param options - The options for the lock manager
   */
  constructor(options: ICellLockManagerOptions) {
    this._doc = options.doc;
    this._permissionManager = options.permissionManager;
    this._defaultTimeout = options.defaultTimeout || 5 * 60 * 1000; // 5 minutes default
    this._cleanupInterval = options.cleanupInterval || 30 * 1000; // 30 seconds default
    
    // Initialize the shared locks map
    this._locks = this._doc.getMap<ICellLock>('cell-locks');
    
    // Set up lock observation
    this._locks.observe(this._onLocksChanged.bind(this));
    
    // Set up cleanup interval
    this._cleanupTimer = setInterval(() => {
      this._cleanupExpiredLocks();
    }, this._cleanupInterval);
  }

  /**
   * Acquire a lock on a cell
   * 
   * @param cellId - The ID of the cell to lock
   * @param userId - The ID of the user acquiring the lock
   * @param userName - The display name of the user
   * @param timeout - Optional custom timeout in milliseconds
   * @returns True if the lock was acquired, false otherwise
   */
  acquireLock(cellId: string, userId: string, userName: string, timeout?: number): boolean {
    // Check if the cell is already locked
    if (this.isLocked(cellId)) {
      const currentLock = this._locks.get(cellId);
      
      // If the same user already has the lock, refresh it
      if (currentLock && currentLock.userId === userId) {
        return this._refreshLock(cellId, userId, userName, timeout);
      }
      
      // Cell is locked by another user
      return false;
    }
    
    // Check if the user has permission to edit this cell
    if (!this._hasEditPermission(userId, cellId)) {
      return false;
    }
    
    // Create the lock
    const lock: ICellLock = {
      cellId,
      userId,
      userName,
      timestamp: Date.now(),
      timeout: timeout !== undefined ? timeout : this._defaultTimeout
    };
    
    // Add the lock to the shared map
    this._doc.transact(() => {
      this._locks.set(cellId, lock);
    });
    
    return true;
  }

  /**
   * Release a lock on a cell
   * 
   * @param cellId - The ID of the cell
   * @param userId - The ID of the user releasing the lock
   * @returns True if the lock was released, false otherwise
   */
  releaseLock(cellId: string, userId: string): boolean {
    const lock = this._locks.get(cellId);
    
    // Check if the cell is locked and the user owns the lock
    if (!lock || lock.userId !== userId) {
      // Check if the user has admin permission to force release
      if (this._hasAdminPermission(userId)) {
        return this.forceReleaseLock(cellId, userId);
      }
      return false;
    }
    
    // Remove the lock
    this._doc.transact(() => {
      this._locks.delete(cellId);
    });
    
    return true;
  }

  /**
   * Force release a lock (admin only)
   * 
   * @param cellId - The ID of the cell
   * @param adminUserId - The ID of the admin user
   * @returns True if the lock was force released, false otherwise
   */
  forceReleaseLock(cellId: string, adminUserId: string): boolean {
    // Check if the user has admin permission
    if (!this._hasAdminPermission(adminUserId)) {
      return false;
    }
    
    const lock = this._locks.get(cellId);
    if (!lock) {
      return false;
    }
    
    // Store the lock for the event
    const lockCopy = { ...lock };
    
    // Remove the lock
    this._doc.transact(() => {
      this._locks.delete(cellId);
    });
    
    // Emit force release event
    this._emitLockEvent({
      type: LockEventType.ForceReleased,
      lock: lockCopy
    });
    
    return true;
  }

  /**
   * Check if a cell is locked
   * 
   * @param cellId - The ID of the cell
   * @returns True if the cell is locked, false otherwise
   */
  isLocked(cellId: string): boolean {
    return this._locks.has(cellId);
  }

  /**
   * Check if a cell is locked by a specific user
   * 
   * @param cellId - The ID of the cell
   * @param userId - The ID of the user
   * @returns True if the cell is locked by the user, false otherwise
   */
  isLockedByUser(cellId: string, userId: string): boolean {
    const lock = this._locks.get(cellId);
    return !!lock && lock.userId === userId;
  }

  /**
   * Get the lock for a cell
   * 
   * @param cellId - The ID of the cell
   * @returns The lock data or null if the cell is not locked
   */
  getLock(cellId: string): ICellLock | null {
    const lock = this._locks.get(cellId);
    return lock || null;
  }

  /**
   * Get all locked cells
   * 
   * @returns An array of all locks
   */
  getAllLocks(): ICellLock[] {
    const locks: ICellLock[] = [];
    this._locks.forEach((lock) => {
      locks.push(lock);
    });
    return locks;
  }

  /**
   * Get all cells locked by a specific user
   * 
   * @param userId - The ID of the user
   * @returns An array of locks owned by the user
   */
  getUserLocks(userId: string): ICellLock[] {
    const locks: ICellLock[] = [];
    this._locks.forEach((lock) => {
      if (lock.userId === userId) {
        locks.push(lock);
      }
    });
    return locks;
  }

  /**
   * Release all locks owned by a user (useful when a user disconnects)
   * 
   * @param userId - The ID of the user
   * @returns The number of locks released
   */
  releaseUserLocks(userId: string): number {
    const userLocks = this.getUserLocks(userId);
    let count = 0;
    
    this._doc.transact(() => {
      userLocks.forEach(lock => {
        this._locks.delete(lock.cellId);
        count++;
      });
    });
    
    return count;
  }

  /**
   * Connect to lock events
   * 
   * @param callback - The callback function to call when a lock event occurs
   * @returns A function to disconnect from lock events
   */
  onLockEvent(callback: (event: ICellLockEvent) => void): () => void {
    return this._lockEvents.connect(callback);
  }

  /**
   * Dispose of the lock manager
   */
  dispose(): void {
    if (this._cleanupTimer) {
      clearInterval(this._cleanupTimer);
      this._cleanupTimer = null;
    }
    
    this._lockEvents.disconnect();
  }

  /**
   * Refresh a lock with a new timestamp
   * 
   * @private
   * @param cellId - The ID of the cell
   * @param userId - The ID of the user
   * @param userName - The display name of the user
   * @param timeout - Optional custom timeout in milliseconds
   * @returns True if the lock was refreshed, false otherwise
   */
  private _refreshLock(cellId: string, userId: string, userName: string, timeout?: number): boolean {
    const currentLock = this._locks.get(cellId);
    if (!currentLock || currentLock.userId !== userId) {
      return false;
    }
    
    // Update the lock with a new timestamp
    const updatedLock: ICellLock = {
      ...currentLock,
      timestamp: Date.now(),
      userName, // Update the name in case it changed
      timeout: timeout !== undefined ? timeout : currentLock.timeout
    };
    
    this._doc.transact(() => {
      this._locks.set(cellId, updatedLock);
    });
    
    return true;
  }

  /**
   * Clean up expired locks
   * 
   * @private
   */
  private _cleanupExpiredLocks(): void {
    const now = Date.now();
    const expiredLocks: ICellLock[] = [];
    
    // Find all expired locks
    this._locks.forEach((lock) => {
      // Skip locks with no timeout
      if (lock.timeout === 0) {
        return;
      }
      
      if (now - lock.timestamp > lock.timeout) {
        expiredLocks.push(lock);
      }
    });
    
    // Remove expired locks
    if (expiredLocks.length > 0) {
      this._doc.transact(() => {
        expiredLocks.forEach(lock => {
          this._locks.delete(lock.cellId);
          
          // Emit timeout event
          this._emitLockEvent({
            type: LockEventType.TimedOut,
            lock
          });
        });
      });
    }
  }

  /**
   * Handle changes to the locks map
   * 
   * @private
   * @param event - The Y.MapEvent
   */
  private _onLocksChanged(event: Y.YMapEvent<ICellLock>): void {
    // Process added or updated locks
    event.keysChanged.forEach(cellId => {
      if (this._locks.has(cellId)) {
        // Lock was added or updated
        const lock = this._locks.get(cellId)!;
        this._emitLockEvent({
          type: LockEventType.Acquired,
          lock
        });
      } else {
        // Lock was removed
        // We need to get the old value from the event
        const oldLock = event.changes.keys.get(cellId)?.oldValue as ICellLock;
        if (oldLock) {
          this._emitLockEvent({
            type: LockEventType.Released,
            lock: oldLock
          });
        }
      }
    });
  }

  /**
   * Emit a lock event
   * 
   * @private
   * @param event - The lock event
   */
  private _emitLockEvent(event: ICellLockEvent): void {
    this._lockEvents.emit(event);
  }

  /**
   * Check if a user has edit permission for a cell
   * 
   * @private
   * @param userId - The ID of the user
   * @param cellId - The ID of the cell
   * @returns True if the user has edit permission, false otherwise
   */
  private _hasEditPermission(userId: string, cellId: string): boolean {
    return this._permissionManager.hasPermission(userId, cellId, UserPermission.Edit);
  }

  /**
   * Check if a user has admin permission
   * 
   * @private
   * @param userId - The ID of the user
   * @returns True if the user has admin permission, false otherwise
   */
  private _hasAdminPermission(userId: string): boolean {
    return this._permissionManager.hasPermission(userId, null, UserPermission.Admin);
  }

  /** The Yjs document */
  private _doc: Y.Doc;
  
  /** The permission manager */
  private _permissionManager: IPermissionManager;
  
  /** The shared locks map */
  private _locks: Y.Map<ICellLock>;
  
  /** Default lock timeout in milliseconds */
  private _defaultTimeout: number;
  
  /** Cleanup interval in milliseconds */
  private _cleanupInterval: number;
  
  /** Timer for cleanup */
  private _cleanupTimer: any;
  
  /** Event emitter for lock events */
  private _lockEvents = new EventEmitter<ICellLockEvent>();
}